### ISSUE TITLE
Add : Soil Type Classification Model

### ISSUE THREAD

**Issue by keshripritesh** [omroy07/AgriTech#424]
### 📌 Project Description

This project is a **Deep Learning-based Soil Type Classification System** that classifies soil images into six categories using **MobileNetV2** architecture. The model helps in **agricultural research, soil analysis, and smart farming applications** by identifying soil types through computer vision techniques.

It can be integrated into a web dashboard or mobile app for farmers, researchers, and agricultural agencies to predict soil type instantly using an image.

---

### 🧠 Tech Stack

* **Language:** Python
* **Libraries/Frameworks:** TensorFlow, Keras, NumPy, OpenCV, Matplotlib
* **Model Architecture:** MobileNetV2 (fine-tuned)
* **Loss Function:** Categorical Crossentropy
* **Optimizer:** Adam
* **Metrics:** Accuracy

---

### 📊 Dataset

* **Source:** [[Kaggle - Soil Types Dataset](https://www.kaggle.com/datasets/matshidiso/soil-types)](https://www.kaggle.com/datasets/matshidiso/soil-types)
* **Classes:**

  * Alluvial soil
  * Clayey soils
  * Laterite soil
  * Loamy soil
  * Sandy loam
  * Sandy soil

**Data Augmentation Techniques:**

* Rotation
* Width/Height shift
* Shear
* Zoom
* Horizontal & vertical flip
* Brightness adjustment

---

### ⚙️ Model Details

* **Base Model:** MobileNetV2 (Pre-trained on ImageNet)
* **Custom Top Layers:** Added dense and dropout layers for fine-tuning
* **Callbacks:**

  * EarlyStopping
  * ReduceLROnPlateau
* **Input Size:** 150x150 pixels
* **Final Validation Set:** 726 images

---

### 📈 Model Performance

| Class         | Precision | Recall | F1-score | Support |
| ------------- | --------- | ------ | -------- | ------- |
| Alluvial soil | 0.75      | 0.33   | 0.46     | 55      |
| Clayey soils  | 1.00      | 0.36   | 0.53     | 144     |
| Laterite soil | 1.00      | 0.36   | 0.53     | 145     |
| Loamy soil    | 0.62      | 0.88   | 0.73     | 97      |
| Sandy loam    | 0.61      | 0.90   | 0.73     | 126     |
| Sandy soil    | 0.57      | 0.99   | 0.72     | 159     |

**Overall Accuracy:** 0.66
**Macro Avg:** Precision = 0.76 | Recall = 0.64 | F1 = 0.62
**Weighted Avg:** Precision = 0.77 | Recall = 0.66 | F1 = 0.63

**Observations:**

* Performs well on Loamy, Sandy loam, and Sandy soil.
* Misclassification mainly between Alluvial and Clayey soils.
* More samples or targeted augmentation could further improve performance.

---

### 🧩 Features

* ✅ Train and evaluate multiple architectures (e.g., EfficientNet, ResNet50).
* ✅ Create a **Streamlit web app** for image-based prediction.
* ✅ Add Grad-CAM visualization for model explainability.
* ✅ Deploy model on **Hugging Face Spaces** or **Render**.
* ✅ Integrate confidence score and prediction history tracking.

---

### 💻 How to Use

```python
from tensorflow.keras.models import load_model
from tensorflow.keras.preprocessing import image
import numpy as np

model = load_model("soil_classification_model.h5")

# Load and preprocess image
img_path = "path_to_new_image.jpg"
img = image.load_img(img_path, target_size=(150, 150))
x = image.img_to_array(img)/255.0
x = np.expand_dims(x, axis=0)

# Predict
pred = model.predict(x)
pred_class = np.argmax(pred)
confidence = pred[0][pred_class]

print(f"Predicted Soil Type: {class_labels[pred_class]}")
print(f"Confidence: {confidence:.2f}")
```

---

### 🏆 Expected Outcomes

* A high-performing deep learning model for soil classification.
* An interactive, user-friendly interface for real-time soil prediction.
* Open-source contributions to enhance agricultural AI applications.

---

### 📂 Repository Setup Tasks for Contributors

* [ ] Add README with project setup and model usage.
* [ ] Add requirements.txt for dependencies.
* [ ] Upload base trained model (optional link).
* [ ] Create Streamlit web app folder (`app/`).
* [ ] Add dataset preprocessing scripts.
* [ ] Add training and evaluation notebooks.

---
